File: scripts/dev-smoke-test.mjs
#!/usr/bin/env node
/**
 * npm 包端到端冒烟测试：验证 CLI 的 install / verify / upgrade / uninstall
 * 到临时目录（v0.2.0 布局：资产在 .drogon-plugin/，项目根文件不受影响）。
 * 用法: node scripts/dev-smoke-test.mjs  （或 npm test）
 */
import { execFileSync } from 'node:child_process'
import { createRequire } from 'node:module'
import { fileURLToPath } from 'node:url'
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

const require = createRequire(import.meta.url)
const PKG = require('../npm/package.json')

// 技能数单一事实源:源仓库 skills/ 目录(不依赖 cwd,禁止硬编码)。
const REPO = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..')
const EXPECTED_SKILLS = fs
  .readdirSync(path.join(REPO, 'skills'), { withFileTypes: true })
  .filter((d) => d.isDirectory()).length

const EXE = path.resolve(
  process.cwd(),
  'node_modules',
  '.bin',
  process.platform === 'win32' ? 'drogon-claude-plugin.cmd' : 'drogon-claude-plugin'
)

function run(args, opts = {}) {
  const bin = fs.existsSync(EXE) ? EXE : 'drogon-claude-plugin'
  const r = (() => {
    try {
      return execFileSync(bin, args, {
        encoding: 'utf-8',
        stdio: ['ignore', 'pipe', 'pipe'],
        ...opts,
      })
    } catch (e) {
      return { status: e.status, stdout: e.stdout, stderr: e.stderr }
    }
  })()
  if (typeof r === 'string') return { status: 0, stdout: r, stderr: '' }
  return r
}

async function main() {
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'drogon-npm-smoke-'))
  const target = path.join(tmp, 'proj')
  fs.mkdirSync(target, { recursive: true })
  const plugin = path.join(target, '.drogon-plugin')

  // 项目自有文件：安装/升级/卸载全程不得触碰
  fs.writeFileSync(path.join(target, 'CLAUDE.md'), '# 我的项目\n', 'utf-8')

  let r = run(['version'])
  if (r.status !== 0) throw new Error(`version 失败: ${r.stderr}`)
  console.log('✓ version')

  r = run(['install', '--target', target])
  if (r.status !== 0) throw new Error(`install 失败: ${r.stderr}`)
  for (const rel of [
    '.claude-plugin/plugin.json',
    '.zcode-plugin/plugin.json',
    'CLAUDE.md',
    'hooks/hooks.json',
    'hooks/run-hook.cmd',
    'hooks/session-start',
    'hooks/post-tool-use',
    'hooks/posttooluse.py',
  ]) {
    if (!fs.existsSync(path.join(plugin, rel)))
      throw new Error(`install 后缺少 ${rel}`)
  }
  const skills = fs
    .readdirSync(path.join(plugin, 'skills'), { withFileTypes: true })
    .filter((d) => d.isDirectory())
  if (skills.length !== EXPECTED_SKILLS)
    throw new Error(`skills 数 ${skills.length} != 源仓库 ${EXPECTED_SKILLS}`)
  if (fs.readFileSync(path.join(target, 'CLAUDE.md'), 'utf-8') !== '# 我的项目\n')
    throw new Error('install 覆盖了项目自有 CLAUDE.md')
  console.log('install OK')

  r = run(['verify', '--target', target])
  if (r.status !== 0 || !String(r.stdout).includes('✅ 通过'))
    throw new Error(`verify 未通过: ${r.stdout} ${r.stderr}`)
  console.log('verify OK')

  r = run(['upgrade', '--target', target])
  if (r.status !== 0 || !String(r.stdout).includes('已是最新'))
    throw new Error(`upgrade(同版本) 未通过: ${r.stdout} ${r.stderr}`)
  console.log('upgrade OK')

  r = run(['uninstall', '--target', target])
  if (r.status !== 0 || fs.existsSync(plugin))
    throw new Error(`uninstall 未清空: ${r.stderr}`)
  if (fs.readFileSync(path.join(target, 'CLAUDE.md'), 'utf-8') !== '# 我的项目\n')
    throw new Error('uninstall 误删了项目自有 CLAUDE.md')
  console.log('uninstall OK')

  fs.rmSync(tmp, { recursive: true, force: true })
  console.log('✅ npm CLI 冒烟测试通过')
}

main().catch((e) => {
  console.error(`❌ ${e.message}`)
  process.exit(1)
})

File: scripts/dev-smoke-test.py
#!/usr/bin/env python3
"""端到端冒烟测试：验证 PyPI CLI 安装/校验/升级/卸载 到临时目录。

用法:
    python scripts/dev-smoke-test.py [--keep]

要求: 先安装本包（pip install -e . 或 .whl）.
所有子进程调用均为显式参数列表 + shell=False，不经过任何 shell。
"""
import argparse
import subprocess
import sys
import tempfile
from pathlib import Path


def run(cli_args):
    """以参数列表运行 drogon-claude-plugin（shell=False，无拼接）。"""
    print(f"$ drogon-claude-plugin {' '.join(cli_args)}")
    r = subprocess.run(
        ["drogon-claude-plugin", *cli_args],
        capture_output=True,
        text=True,
        shell=False,
    )
    print(r.stdout, end="")
    if r.stderr:
        print(r.stderr, file=sys.stderr, end="")
    return r.returncode, r.stdout


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--keep", action="store_true")
    args = ap.parse_args()

    tmp = Path(tempfile.mkdtemp(prefix="drogon-smoke-"))
    target = tmp / "proj"
    target.mkdir()
    plugin = target / ".drogon-plugin"

    # 项目自有文件：安装/升级/卸载全程不得触碰
    own_md = target / "CLAUDE.md"
    own_md.write_text("# 我的项目\n", encoding="utf-8")

    # version
    rc, _ = run(["version"])
    if rc != 0:
        return 1

    # install → .drogon-plugin/
    rc, _ = run(["install", "--target", str(target)])
    if rc != 0:
        return 1
    for rel in (
        ".claude-plugin/plugin.json",
        ".zcode-plugin/plugin.json",
        "CLAUDE.md",
        "hooks/hooks.json",
        "hooks/run-hook.cmd",
        "hooks/session-start",
        "hooks/post-tool-use",
        "hooks/posttooluse.py",
    ):
        if not (plugin / rel).is_file():
            print(f"❌ 缺少 {rel}")
            return 1
    n_skills = len(list((plugin / "skills").iterdir()))
    if n_skills != 22:
        print(f"❌ skills 数 {n_skills} != 22")
        return 1
    if own_md.read_text(encoding="utf-8") != "# 我的项目\n":
        print("❌ 安装覆盖了项目自有 CLAUDE.md")
        return 1

    # verify
    rc, out = run(["verify", "--target", str(target)])
    if rc != 0 or "✅ 通过" not in out:
        print("❌ verify 未通过")
        return 1

    # upgrade（同版本 → 提示已是最新）
    rc, out = run(["upgrade", "--target", str(target)])
    if rc != 0 or "已是最新" not in out:
        print("❌ upgrade(同版本) 未通过")
        return 1

    # uninstall
    rc, _ = run(["uninstall", "--target", str(target)])
    if rc != 0 or plugin.exists():
        print("❌ uninstall 未清空")
        return 1
    if own_md.read_text(encoding="utf-8") != "# 我的项目\n":
        print("❌ 卸载误删了项目自有 CLAUDE.md")
        return 1

    print("✅ PyPI CLI 冒烟测试通过")
    if not args.keep:
        import shutil

        shutil.rmtree(tmp, ignore_errors=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())
